codegen/main.go

```diff
@@ -0,0 +1,461 @@
+	IsDeprecated       bool
+	DeprecationWarning string
+	IsHidden           bool
+	IsShorthandHidden  bool
+}
+
+type templateData struct {
+	TypeTemplateData []typeTemplateData
+	FlagTemplateData []flagTemplateData
+	// Back-ticks are not supported in templates. So, passing as a parameter.
+	BT string
+}
+
+type Param struct {
+	FlagName           string `yaml:"flag-name"`
+	Shorthand          string
+	Type               string
+	DefaultValue       string `yaml:"default"`
+	ConfigPath         string `yaml:"config-path"`
+	IsDeprecated       bool   `yaml:"deprecated"`
+	DeprecationWarning string `yaml:"deprecation-warning"`
+	Usage              string
+	HideFlag           bool `yaml:"hide-flag"`
+	HideShorthand      bool `yaml:"hide-shorthand"`
+}
+
+func capitalizeIdentifier(name string) (string, error) {
+	if !paramNameRegexCompiled.MatchString(name) {
+		return "", fmt.Errorf("%s is not a supported name", name)
+	}
+	id := ""
+	capitalizeNextRune := true
+	for _, v := range name {
+		if v != '-' && v != '.' {
+			if capitalizeNextRune {
+				v = unicode.ToUpper(v)
+			}
+			id = id + string(v)
+			capitalizeNextRune = false
+		} else {
+			capitalizeNextRune = true
+		}
+	}
+	return id, nil
+}
+
+type typeInfo struct {
+	typeName   string
+	fieldName  string
+	configPath string
+	dataType   string
+}
+
+func getGoDataType(dt string) (string, error) {
+	switch dt {
+	case "octal":
+		return "Octal", nil
+	case "duration":
+		return "time.Duration", nil
+	default:
+		return dt, nil
+	}
+}
+
+// Figures out what all structs and corresponding fields need to be created for
+// a given param. Other params can add fields to the same type if they share the
+// same config-path hierarchy.
+// For a config-path of x.y for an int param it will yield the following structs
+//
+//	type Config struct {
+//	  X XConfig `yaml:"x"`
+//	}
+//
+//	type XConfig struct {
+//	  Y int `yaml:"y"`
+//	}
+func typesToCreate(param Param) ([]typeInfo, error) {
+	configPath := param.ConfigPath
+	typeNameSuffix := "Config"
+	typesFieldPairs := make([]typeInfo, 0)
+	for configPath != "" {
+		nestedTypeName := ""
+		nextItrCfgPath := ""
+		curCfgPath := configPath
+		dataType, err := getGoDataType(param.Type)
+		if err != nil {
+			return nil, err
+		}
+		dotIdx := strings.IndexRune(configPath, '.')
+		if dotIdx != -1 {
+			// Config path contains at least one dot which means it's not a leaf element.
+			curCfgPath = configPath[:dotIdx]
+			nextItrCfgPath = configPath[dotIdx+1:]
+			nestedTypeName, err = capitalizeIdentifier(curCfgPath)
+			if err != nil {
+				return nil, err
+			}
+			dataType = nestedTypeName + typeNameSuffix
+		}
+		fldName, err := capitalizeIdentifier(curCfgPath)
+		if err != nil {
+			return nil, err
+		}
+		typesFieldPairs = append(typesFieldPairs, typeInfo{
+			typeName:   typeNameSuffix,
+			fieldName:  fldName,
+			dataType:   dataType,
+			configPath: curCfgPath,
+		})
+		if err != nil {
+			return nil, err
+		}
+		typeNameSuffix = dataType
```

Comment:
Shouldn't this always be "Config"

For a config-path of x.y.z for an int param it will yield the following structs
//
//	type Config struct {
//	  X XConfig `yaml:"x"`
//	}
//
//	type XConfig struct {
//	  Y YConfig `yaml:"y"`
//	}
//
//	type ZConfig struct {
//	  Z int `yaml:"z"`
//	}

Reply:
No, a config path of x.y.z might conflict with y.z even though they have different hierarchies. To prevent this, we generate the following structs
type Config struct {
	X XConfig `yaml:"x"`

	Y YConfig `yaml:"y"`
}

type XConfig struct {
	Y YXConfig `yaml:"y"`
}

type YConfig struct {
	Z string `yaml:"z"`
}

type YXConfig struct {
	Z string `yaml:"z"`
}